src/parser/on_demand/parse_to_elements.ts

```diff
@@ -0,0 +1,131 @@
+import { BSONOffsetError } from '../../error';
+import { NumberUtils } from '../../utils/number_utils';
+
+/**
+ * @public
+ * @experimental
+ */
+export type BSONElement = [
+  type: number,
+  nameOffset: number,
+  nameLength: number,
+  offset: number,
+  length: number
+];
+
+/**
+ * Searches for null terminator.
+ * **Does not** bounds check since this should **ONLY** be used within parseToElements which has asserted that `bytes` ends with a `0x00`.
+ * So this will at most iterate to the document's terminator and error if that is the offset reached.
+ */
+function findNull(bytes: Uint8Array, offset: number): number {
+  let nullTerminatorOffset = offset;
+
+  for (; bytes[nullTerminatorOffset] !== 0x00; nullTerminatorOffset++);
+
+  if (nullTerminatorOffset === bytes.length - 1) {
```

Comment:
Is it necessary to throw?  The doc comment says this is only used in places that have already asserted there's a null terminator, so we will never enter this if-statement

Reply:
Yes, added a comment to clarify. We don't need to bounds check the iteration (as in `bytes[nullTerminatorOffset] !== 0x00 && bytes.length < nullTerminatorOffset`) since we know we will always encounter a null byte at the end of the array, so we won't loop forever. However, if the null byte we reach is the last one in the array then it is not a null byte for a value rather the one for the document, so we've found "no null byte (for the value)"